### DISCUSSION TITLE
OPC UA Server Implementation

### DISCUSSION THREAD

**Discussion by TelmoSantosCarneiro** [gopcua/opcua#659] [Q&A] [answered]
Will we be able to implement an OPC UA Server using this library in the near future? Could you tell me if there is anything planned in this regard? I have seen https://github.com/gopcua/opcua/discussions/429#discussion-3311414, is there any updates?

Thanks

**Comment by magiconair** [comment]
It depends on whether I can get the energy to lay the foundation. We need an address space implementation and a good architecture for implementing missing functionality. So far this topic has come up every now and then but I don't know how much desire there is. At least the branch should be up-to-date. Let me try with another push. I think once we have the basic functionality there people can provide PRs to fill the gaps.

**Comment by magiconair** [comment]
OK, so I've picked this up again and made some progress. Expect the `server` branch to be a bit of a wild ride while I'm getting my bearings. I've refactored the address space so that you can now define a user address space and define nodes with dynamic values. Also fixed a bug which would prevent the ProSys client to connect. 

It should be obvious that this is not production ready and expect things to change quite a bit but if you are interested in playing with the server and telling me where it works and it doesn't then this looks like a good time since there is some momentum.

**Comment by magiconair** [comment]
Check `examples/server` for how to use the user address space.

**Comment by magiconair** [comment]
@kung-foo just FYI :)

**Comment by kung-foo** [comment]
We're getting pretty close at Intelecy to either directly funding, or at least allocating a resource to work on the server implementation. Hopefully just after summer (since no one works in the Nordics when the sun is out) ;)

**Comment by magiconair** [comment]
That would be great. I am making some progress in getting this to work. The ProSys client connects properly. Now I'm stuck with browsing and that's a big can of worms. But I think once that is in place we should have a pretty solid foundation.

**Comment by KebSeb** [comment]
Hi,
Do we have progress here? I woul dalso be very interested in the Server

**Reply by magiconair** [reply]
Yes, there is some progress but I'm currently struggling to implement Browse correctly and twist my brain on following the reference tree. It is in the server branch but nothing you can use that and the implementation still changes as I'm building things.

**Comment by danomagnum** [comment]
There is now an experimental server release.  It supports most of the common data access services (browse, read, write, subscribe, etc...).

https://github.com/gopcua/opcua/releases/tag/v0.5.3-server

**Comment by danomagnum** [comment, marked as answer]
Server is has been in since version 0.6.0
